wasi/wasi_test.go

```diff
@@ -0,0 +1,199 @@
+
+			// Compare them
+			errno := wasiEnv.args_sizes_get(hostFunctionCallContext, argCountPtr, bufSizePtr)
+			require.Equal(t, ESUCCESS, errno)
+			require.Equal(t, expectedArgCount, hostFunctionCallContext.Memory.Buffer[argCountPtr:argCountPtr+4])
+			require.Equal(t, expectedBufSize, hostFunctionCallContext.Memory.Buffer[bufSizePtr:bufSizePtr+4])
+
+			// Serialize the expected result of args_get
+			expectedArgs := make([]byte, 4*len(tc.args))
+			argsPtr := uint32(0)
+			expectedArgv := make([]byte, tc.expectedBufSize)
+			argvPtr := uint32(0x100)
+			argvWritten := uint32(0)
+			for i, arg := range tc.expectedArgs {
+				binary.LittleEndian.PutUint32(expectedArgs[argsPtr+uint32(i*4):], argvPtr+argvWritten)
+				copy(expectedArgv[argvWritten:], arg)
+				argvWritten += uint32(len(arg))
+			}
+
+			// Compare them
+			errno = wasiEnv.args_get(hostFunctionCallContext, argsPtr, argvPtr)
+			require.Equal(t, ESUCCESS, errno)
+			require.Equal(t, expectedArgs, hostFunctionCallContext.Memory.Buffer[argsPtr:argsPtr+uint32(len(expectedArgs))])
+			require.Equal(t, expectedArgv, hostFunctionCallContext.Memory.Buffer[argvPtr:argvPtr+uint32(len(expectedArgv))])
+		})
+	}
+}
+
+func TestArgsAPIReturnError(t *testing.T) {
+	dummyArgs := []string{"foo", "bar", "baz"}
+	argsOpt, err := Args(dummyArgs)
+	require.NoError(t, err)
+	wasiEnv := NewEnvironment(argsOpt)
+	hostFunctionCallContext := buildMockHostFunctionCallContext()
+
+	outOfBounds := uint32(len(hostFunctionCallContext.Memory.Buffer))
+
+	argsArray, err := newWASIStringArray(dummyArgs)
+	require.NoError(t, err)
+
+	errno := wasiEnv.args_sizes_get(hostFunctionCallContext, outOfBounds, 0)
+	require.Equal(t, EINVAL, errno)
+	errno = wasiEnv.args_sizes_get(hostFunctionCallContext, 0, outOfBounds)
+	require.Equal(t, EINVAL, errno)
+	maxValidArgsCountPtr := outOfBounds - 4
+	errno = wasiEnv.args_sizes_get(hostFunctionCallContext, maxValidArgsCountPtr+1, 0)
+	require.Equal(t, EINVAL, errno)
+	maxValidBufSizeCountPtr := outOfBounds - 4
+	errno = wasiEnv.args_sizes_get(hostFunctionCallContext, 0, maxValidBufSizeCountPtr+1)
+	require.Equal(t, EINVAL, errno)
+
+	errno = wasiEnv.args_get(hostFunctionCallContext, outOfBounds, 0)
+	require.Equal(t, EINVAL, errno)
+	errno = wasiEnv.args_get(hostFunctionCallContext, 0, outOfBounds)
+	require.Equal(t, EINVAL, errno)
+	maxValidArgsPtr := outOfBounds - 4*uint32(len(argsArray.strings))
+	errno = wasiEnv.args_get(hostFunctionCallContext, maxValidArgsPtr+1, 0)
+	require.Equal(t, EINVAL, errno)
+	maxValidArgsBufPtr := outOfBounds - argsArray.totalBufSize
+	errno = wasiEnv.args_get(hostFunctionCallContext, 0, maxValidArgsBufPtr+1)
+	require.Equal(t, EINVAL, errno)
+}
+
+func buildMockHostFunctionCallContext() *wasm.HostFunctionCallContext {
+	return &wasm.HostFunctionCallContext{
+		Memory: &wasm.MemoryInstance{Buffer: make([]byte, wasm.PageSize), Min: 1},
+	}
+}
+
+// Test whether the specs implemented by the args API match those understood by a language runtime.
```

Comment:
This test isn't a unit test, it is closest to `interop` for lack of them. That this keeps coming up is important, but it is also important to not add more tinygo deps to the main tree just for lack of a better directory. The best time to solve this is now,

Can you please make a `interop/wasi` root directory and move this here. Inside make a test for args_get and args_sizes_get with the pre-compiled wasm in `interop/wasi/testdata`. Make can refresh that from tinygo files also in that directory.

Ex you can start a file `interop/wasi/wasi_interop_tests.go`:
```
// package wasi_interop tests wazero's WASI XX_spec_draft_XXX is compatible with at least one runtime.
//
// We use tinygo, to perform these tests as the maintainers are familiar with it.
//
// Note: Use `make XXX` to compile `testdata/*.wasm` from `testdata/*.go` 
// Note: This also substitutes for WASI spec tests until we have another option
package wasi_interop

// argsGetBinary is compiled from testdata/args_get.go
//go:embed testdata/args_get.wasm
var argsGetBinary []byte
func Test_ArgsGet(t *testing.T) {
```

Reply:
Moved them. Thanks for the feedback!

Reply:
Maybe I'd prefer the name `tests` directory rather than `interop` where I plan to move `wasm/internal/spectests` into `tests/spectests` and will add bunch of test modules under `tests/wasm` which I will use to find compiler bugs

Reply:
if by "tests" you mean "tests/interop" works for me. I would like to isolate where tinygo is required as we can centralize the README and also make can invalidate easier when things are in the same place like that.

Reply:
Ah I see! sg!

Reply:
so I've created the top `tests` directory, feel free to put `interop` under there

Reply:
I've moved to them `tests/wasi` as we discussed offline. Thanks all.